edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/InvokeMethodHandler.cs

```diff
@@ -0,0 +1,157 @@
+
+    public class InvokeMethodHandler : IInvokeMethodHandler
+    {
+        readonly ConcurrentDictionary<string, ConcurrentDictionary<DirectMethodRequest, TaskCompletionSource<DirectMethodResponse>>> clientMethodRequestQueue = new ConcurrentDictionary<string, ConcurrentDictionary<DirectMethodRequest, TaskCompletionSource<DirectMethodResponse>>>();
+        readonly IConnectionManager connectionManager;
+
+        public InvokeMethodHandler(IConnectionManager connectionManager)
+        {
+            this.connectionManager = Preconditions.CheckNotNull(connectionManager, nameof(connectionManager));
+        }
+
+        public Task<DirectMethodResponse> InvokeMethod(DirectMethodRequest methodRequest)
+        {
+            Preconditions.CheckNotNull(methodRequest, nameof(methodRequest));
+
+            Option<IDeviceProxy> deviceProxy = this.GetDeviceProxyWithSubscription(methodRequest.Id);
+            return deviceProxy.Map(
+                    d =>
+                    {
+                        Events.InvokingMethod(methodRequest);
+                        return d.InvokeMethodAsync(methodRequest);
+                    })
+                .GetOrElse(
+                    async () =>
+                    {
+                        var taskCompletion = new TaskCompletionSource<DirectMethodResponse>();
+                        ConcurrentDictionary<DirectMethodRequest, TaskCompletionSource<DirectMethodResponse>> clientQueue = this.GetClientQueue(methodRequest.Id);
+                        clientQueue.TryAdd(methodRequest, taskCompletion);
+                        Task completedTask = await Task.WhenAny(taskCompletion.Task, Task.Delay(methodRequest.ConnectTimeout));
+                        if (completedTask != taskCompletion.Task && clientQueue.TryRemove(methodRequest, out taskCompletion))
+                        {
+                            taskCompletion.TrySetResult(new DirectMethodResponse(new EdgeHubTimeoutException($"Client {methodRequest.Id} not found"), HttpStatusCode.NotFound));
+                        }
+
+                        return await taskCompletion.Task;
+                    });
+        }
+
+        public Task ProcessInvokeMethodSubscription(string id)
+        {
+            this.ProcessInvokeMethodSubscriptionInternal(id);
+            return Task.CompletedTask;
+        }
+
+        async void ProcessInvokeMethodSubscriptionInternal(string id)
+        {
+            try
+            {
+                Events.ProcessingInvokeMethodQueue(id);
+                // Temporary hack to wait for the subscription call to complete. Without this,
+                // the EdgeHub will invoke the pending method request "too soon", before the layers
+                // in between have been set up correctly. To fix this, changes are needed in ProtocolGateway,
+                // Client SDK, and need to figure out a way to raise events for AMQP Links. 
+                await Task.Delay(TimeSpan.FromSeconds(3));
```

Comment:
Changes are required in Client SDK too? Curious, what those are.

Reply:
I suppose we basically want to delay this till an MQTT topic subscription or an AMQP link is setup for the method call no?

Reply:
If we must put in this hack, would it be better to have a loop that periodically polls to see if we get something other than `None` from `GetDeviceProxyWIthSubscription`? Maybe it can timeout after a while.